test/hotspot/jtreg/compiler/loopopts/superword/MinMaxRed_Long.java

```diff
@@ -0,0 +1,155 @@
+                           "-XX:LoopUnrollLimit=250",
+                           "-XX:CompileThresholdScaling=0.1");
+        framework.start();
+    }
+
+    @Run(test = {"maxReductionImplement"},
+         mode = RunMode.STANDALONE)
+    public void runMaxTest() {
+        runMaxTest(50);
+        runMaxTest(80);
+        runMaxTest(100);
+    }
+
+    private static void runMaxTest(int probability) {
+        long[] longs = new long[1024];
+        ReductionInit(longs, probability);
+        long res = 0;
+        for (int j = 0; j < 2000; j++) {
+            res = maxReductionImplement(longs, res);
+        }
+        if (res == 11 * Arrays.stream(longs).max().getAsLong()) {
+            System.out.println("Success");
+        } else {
+            throw new AssertionError("Failed");
+        }
+    }
+
+    @Run(test = {"minReductionImplement"},
+         mode = RunMode.STANDALONE)
+    public void runMinTest() {
+        runMinTest(50);
+        runMinTest(80);
+        runMinTest(100);
+    }
+
+    private static void runMinTest(int probability) {
+        long[] longs = new long[1024];
+        ReductionInit(longs, probability);
+        // Negating the values generated for controlling max branching
+        // allows same logic to be used for min tests.
+        longs = negate(longs);
+        long res = 0;
+        for (int j = 0; j < 2000; j++) {
+            res = minReductionImplement(longs, res);
+        }
+        if (res == 11 * Arrays.stream(longs).min().getAsLong()) {
+            System.out.println("Success");
+        } else {
+            throw new AssertionError("Failed");
+        }
+    }
+
+    static long[] negate(long[] nums) {
+        return LongStream.of(nums).map(l -> -l).toArray();
+    }
+
+    public static void ReductionInit(long[] longs, int probability) {
+        int aboveCount, abovePercent;
+
+        // Iterate until you find a set that matches the requirement probability
+        do {
+            long max = random.nextLong(10);
+            longs[0] = max;
+
+            aboveCount = 0;
+            for (int i = 1; i < longs.length; i++) {
+                long value;
+                if (random.nextLong(101) <= probability) {
+                    long increment = random.nextLong(10);
+                    value = max + increment;
+                    aboveCount++;
+                } else {
+                    // Decrement by at least 1
+                    long decrement = random.nextLong(10) + 1;
+                    value = max - decrement;
+                }
+                longs[i] = value;
+                max = Math.max(max, value);
+            }
+
+            abovePercent = ((aboveCount + 1) * 100) / longs.length;
+        } while (abovePercent != probability);
+    }
+
+    @Test
+    @IR(applyIf = {"SuperWordReductions", "true"},
+        applyIfCPUFeatureOr = { "avx512", "true" },
+        counts = {IRNode.MIN_REDUCTION_V, " > 0"})
```

Comment:
> @eme64 I've addressed all your comments except aarch64 testing. `asimd` is not enough, you need `sve` for this, but I'm yet to make it work even with `sve`, something's up and need to debug it further.

Hi @galderz , may I ask if these long-reduction cases can't work even with `sve`? It might be related with the limitation [here](https://github.com/openjdk/jdk/blob/75420e9314c54adc5b45f9b274a87af54dd6b5a8/src/hotspot/share/opto/superword.cpp#L1564-L1566). Some `sve` machines have only 128 bits.

Reply:
That's right. Neoverse V2 is 4 pipes of 128 bits, V1 is 2 pipes of 256 bits.
That comment is "interesting". Maybe it should be tunable by the back end. Given that Neoverse V2 can issue 4 SVE operations per clock cycle, it might still be a win.

Galder, how about you disable that line and give it another try?